compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/ExpressionCodegen.kt

```diff
@@ -60,110 +52,211 @@ import org.jetbrains.org.objectweb.asm.Type
+        infos.add(info)
+        try {
+            return f(info)
+        } finally {
+            infos.pop()
+        }
     }
 
-    fun removeInfo(info: ExpressionInfo) {
-        assert(peek() == info)
-        pop()
+    inline fun <R> handleBlock(f: (ExpressionInfo) -> R): R? {
+        if (infos.isEmpty()) {
+            return null
+        }
+        val top = infos.pop()
+        try {
+            return f(top)
+        } finally {
+            infos.add(top)
+        }
     }
 
-    fun pop(): ExpressionInfo = infos.pop()
+    companion object {
+        fun create() = BlockInfo(null)
+    }
+}
 
-    fun peek(): ExpressionInfo = infos.peek()
+class VariableInfo(val declaration: IrVariable, val index: Int, val type: Type, val startLabel: Label)
 
-    fun isEmpty(): Boolean = infos.isEmpty()
+// A value that may not have been fully constructed yet. The ability to "roll back" code generation
+// is useful for certain optimizations.
+abstract class PromisedValue(val codegen: ExpressionCodegen, val type: Type) {
+    // If this value is immaterial, construct an object on the top of the stack. This
+    // must always be done before generating other values or emitting raw bytecode.
+    abstract fun materialize()
+}
 
-    fun hasFinallyBlocks(): Boolean = infos.firstIsInstanceOrNull<TryInfo>() != null
+// A value that *has* been fully constructed.
+class MaterialValue(codegen: ExpressionCodegen, type: Type) : PromisedValue(codegen, type) {
+    override fun materialize() {}
+}
 
-    companion object {
-        fun create() = BlockInfo(null)
+// A value that can be branched on. JVM has certain branching instructions which can be used
+// to optimize these.
+abstract class BooleanValue(codegen: ExpressionCodegen) : PromisedValue(codegen, Type.BOOLEAN_TYPE) {
+    abstract fun jumpIfFalse(target: Label)
+    abstract fun jumpIfTrue(target: Label)
+
+    override fun materialize() {
+        val const0 = Label()
+        val end = Label()
+        jumpIfFalse(const0)
+        codegen.mv.iconst(1)
+        codegen.mv.goTo(end)
+        codegen.mv.mark(const0)
+        codegen.mv.iconst(0)
+        codegen.mv.mark(end)
     }
 }
 
-class VariableInfo(val declaration: IrVariable, val index: Int, val type: Type, val startLabel: Label)
+// Same as materialize(), but return a representation of the result.
+val PromisedValue.materialized: MaterialValue
+    get() {
+        materialize()
+        return MaterialValue(codegen, type)
+    }
+
+// Materialize and disregard this value. Materialization is forced because, presumably,
+// we only wanted the side effects anyway.
+fun PromisedValue.discard(): MaterialValue {
+    materialize()
+    if (type !== Type.VOID_TYPE)
+        AsmUtil.pop(codegen.mv, type)
+    return MaterialValue(codegen, Type.VOID_TYPE)
+}
+
+// On materialization, cast the value to a different type.
+fun PromisedValue.coerce(target: Type) = when (target) {
+    Type.VOID_TYPE -> discard()
```

Comment:
Why are we materializing the value if the expected type is void, and don't materialize it otherwise? This doesn't look very obvious

Reply:
This branch was probably necessary to avoid breaking something while I was iterating on the code, but doesn't seem to affect anything now, so I'll remove it.